pages/advanced-concepts/optimisations/cairo-profiler.md

```diff
@@ -0,0 +1,282 @@
+Storage Reads: 1
+Storage Writes: 101
+```
+
+## Advanced Profiling Features
+
+1. **Inlined Function Analysis**
+
+Enable with `--show-inlined-functions` to see detailed call stacks including inlined functions:
+
+```cairo
+cairo-profiler trace.bin --show-inlined-functions
+```
+
+2. **Generic Function Analysis**
+
+Use `--split-generics` to analyze generic functions by their concrete types:
+
+```cairo
+cairo-profiler trace.bin --split-generics
+```
+
+3. **Custom Resource Costs**
+
+Specify custom resource costs for different Starknet versions:
+
+```cairo
+cairo-profiler trace.bin --versioned-constants-path custom_costs.json
+```
+
+## Best Practices
+
+1. **Storage Pattern Optimization**
+
+- Cache storage reads in memory when accessed multiple times
+- Batch storage writes where possible
+- Use storage spans for array operations
+
+2. **Function Call Optimization**
+
+- Monitor inlined vs non-inlined function costs
+- Use `--show-inlined-functions` to identify expensive inlined operations
+- Consider function composition based on profiler output
+
+3. **Memory Management**
+
+- Track memory_holes metric to identify inefficient memory usage
+- Use appropriate data structures to minimize memory fragmentation
+- Consider span operations for array access
+
+4. **L1 Communication Optimization**
+
+- Monitor l2_l1_message_sizes for cross-layer communication
+- Batch L1 messages when possible
+- Compress data when sending to L1
+
+## Common Optimization Patterns
+
+1. **Storage Read Caching**
+
+```cairo
+// Before
+let value1 = storage.read(key);
+let value2 = storage.read(key);
+
+// After
+let cached = storage.read(key);
+let value1 = cached;
+let value2 = cached;
+```
+
+2. **Batch Operations**
+
+```cairo
+// Before
+for item in items {
+    storage.write(item.key, item.value);
+}
+
+// After
+let mut writes = Array::new();
+// ... collect writes ...
+storage.write_multiple(writes);
+```
+
+3. **Memory Efficiency**
+
+```cairo
+// Before
+let mut array = ArrayTrait::new();
+// ... multiple push operations ...
+
+// After
+let mut array = ArrayTrait::with_capacity(size);
+// ... reserve space first ...
+```
+
+## Profile-Guided Optimization Process
+
+1. **Baseline Profiling**
+
+```cairo
+cairo-profiler initial_trace.bin --output-path baseline.pb.gz
+```
+
+2. **Identify Hotspots**
+
+```cairo
+Use flame graphs to identify expensive operations
+Focus on functions with high step counts
+Look for repeated storage operations
+```
+
+3. **Optimize and Compare**
+
+```cairo
+cairo-profiler optimized_trace.bin --output-path optimized.pb.gz
+```
+
+4. **Validate Improvements**
+
+- Compare step counts and resource usage
+- Verify optimization doesn't introduce new inefficiencies
+- Document performance gains
+
+## Resources and Tools
+
+1. **pprof visualization**:
+
+```cairo
+go tool pprof -http=:8080 profile.pb.gz
+```
+
+2. **Flame graph generation**:
+
+```cairo
+cairo-profiler trace.bin --output-path flame.svg --format=flamegraph
+```
```

Comment:
Can you add screenshot of pprof graphs on the Unoptimized and Optimized contracts. You can move this up a bit